backend/rules_engine.py

```diff
@@ -0,0 +1,42 @@
+import logging
+from typing import List, Dict, Callable
+
+from config import Rule
+from models import Transaction
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+logger.addHandler(logging.StreamHandler())
+
+class RulesEngine:
+
+    """
+    Given a list of rules and a transactions, this method updates the transaction category_id if it matches
+    a rule.  This function stops when it finds the first rule that matches the transaction.  If no rules match, then
+    the transaction category_id is not changed.
+    """
+    @staticmethod
+    def apply_rules(rules: List[Rule], transaction: Transaction) -> None:
+        for rule in rules:
+            if RulesEngine.is_match(transaction.payee, rule):
+                transaction.category_id = rule.category_id
+                break  # Stop on the first matching rule.
+
+    @staticmethod
+    def is_match(payee: str, rule: Rule) -> bool:
+        matching_functions: Dict[Rule.MatchingOperator, Callable[[str], bool]] = {
```

Comment:
Weigh the pros/cons of making this dict a class field so that it's not instantiated each time this method is called, because this method is called many times.